### PR TITLE
Make directories excludable via configuration.

### PR DESCRIPTION
**Pull request by troessner**
Relates to #548 #349 #408 and #529 

Apart from that I hate ruby. I've been doing ruby for years now and when i checked the docs here: http://docs.ruby-lang.org/en/2.1.0/Find.html I thought the 

```Ruby
if File.basename(path)[0] == ?.
```

was a syntax error. Of course it isn't in Ruby.

```Bash
[1] pry(main)> ?a
=> "a"
[2] pry(main)> ?ab
SyntaxError: unexpected '?', expecting end-of-input
```
The dark corners of ruby baffle me every time.

### CODE REVIEW

File: spec/reek/configuration/app_configuration_spec.rb [hunk 1]
```diff
@@ -46,20 +46,33 @@
...
     it 'resets the configuration' do
-      Reek::Configuration::AppConfiguration.load_from_file(sample_configuration_path)
+      described_class.load_from_file(sample_configuration_path)
 
-      expect(Reek::Configuration::AppConfiguration.configuration).to eq(sample_configuration_loaded)
-      Reek::Configuration::AppConfiguration.reset
-      expect(Reek::Configuration::AppConfiguration.configuration).to eq(default_configuration)
+      expect(described_class.configuration).to eq(sample_configuration_loaded)
+      described_class.reset
+      expect(described_class.configuration).to eq(default_configuration)
```
Comment (troessner) [inline spec/reek/configuration/app_configuration_spec.rb]: Not really part of this PR but cleaned up according to the good old boyscout rule.:)

File: lib/reek/source/source_locator.rb [hunk 2]
```diff
@@ -16,22 +18,37 @@ def initialize(paths)
...
-            find_sources(Dir["#{pathname}/**/*.rb"])
-          else
-            next pathname if pathname.file?
-            $stderr.puts "Error: No such file - #{pathname}"
-            nil
+      def source_paths
+        relevant_paths = []
+        @paths.map do |given_path|
+          print_no_such_file_error(given_path) && next unless path_exists?(given_path)
+          Find.find(given_path) do |path|
```
Comment (troessner) [inline lib/reek/source/source_locator.rb]: Didn't know that thing before this PR but it seems very convenient: http://ruby-doc.org/stdlib-2.2.0/libdoc/find/rdoc/Find.html

File: lib/reek/cli/option_interpreter.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@ module CLI
     # Interprets the options set from the command line
     #
     class OptionInterpreter
-      include CLI::Input
```
Comment (troessner) [inline lib/reek/cli/option_interpreter.rb]: Boyscout rule.

File: spec/reek/source/source_locator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+          'spec/samples/source_with_exclude_paths/ignore_me/uncommunicative_method_name.rb',
+          'spec/samples/source_with_exclude_paths/nested/ignore_me_as_well/irresponsible_module.rb'
+        ]
+      end
+
+      it 'does not use excluded paths' do
+        with_test_config(config) do
+          sources = described_class.new([path]).sources
+
+          expect(sources.map(&:description).sort).
```
Comment (troessner) [inline spec/reek/source/source_locator_spec.rb]: I know that this is kind of implicitly checked by the expectation below but this makes it more palpable for the innocent reader.

**Comment by troessner** [comment]
This bad boy is ready for review @mvz && @chastell !

**Comment by chastell** [comment]
Nice!

Every time I encounter the ‘path as `String`’ vs ‘path as `Pathname`’ duality I cringe. Given that `Pathnames` try to be very `String`-like, is there a chance that we try to use `Pathnames` everywhere and `#to_s` them only where needed? By this I mean `SourceLocator#paths` being an `Array` of `Pathnames` (not `Strings`) and `Configuration::AppConfiguration.exclude_paths` being an `Array` of `Pathnames`. If not, then maybe `Configuration::AppConfiguration.excludes_path?` that can take a `Pathname`? This way e.g. `SourceLocator#path_exists?` could be inlined, and, in general, we could start assuming that we have `Pathanames` by default everywhere in `reek`.

File: lib/reek/source/source_locator.rb (hunk 2 above)
Comment (chastell) [inline lib/reek/source/source_locator.rb]: Yes, it’s nice, but is `Find.prune` thread-safe (it uses `catch`/`throw` under the hood)?

(Also, are we interested in `reek` being thread-safe?)

**Comment by troessner** [comment]
>> Every time I encounter the ‘path as String’ vs ‘path as Pathname’ duality I cringe. Given that Pathnames try to be very String-like, is there a chance that we try to use Pathnames everywhere and #to_s them only where needed? By this I mean SourceLocator#paths being an Array of Pathnames (not Strings) and Configuration::AppConfiguration.exclude_paths being an Array of Pathnames. If not, then maybe Configuration::AppConfiguration.excludes_path? that can take a Pathname? This way e.g. SourceLocator#path_exists? could be inlined, and, in general, we could start assuming that we have Pathanames by default everywhere in reek.

I could not agree more. So let's switch to Pathname everywhere.
However in order to reduce complexity I'd prefer to do this in a follow up pull request, and not in this one.

File: lib/reek/source/source_locator.rb (hunk 2 above)
Comment (troessner) [inline lib/reek/source/source_locator.rb]: >> (Also, are we interested in reek being thread-safe?)

Good question.
So what could be the worst thing that happens?
Somebody using reek via our API in 2 different threads with one thread f***ing up "Find" in the other thread?

Comment (chastell) [inline lib/reek/source/source_locator.rb]: Somebody using `reek` in one thread and _anything else_ that uses `Find.prune` in another thread. My very quick-and-dirty proof of concept:

```Ruby
def prune
  throw :prune
end

a = Thread.new do
  puts 'a'
  prune
end

b = Thread.new do
  puts 'b'
  catch(:prune) { puts 'caught' }
end

a.join
b.join
```

– if you run this enough times you’ll see `caught` being printed.

That said, I’m not sure whether `reek` is anywhere to being thread-safe everywhere else; I don’t want this to block this PR as long as we track the knowledge somewhere and come back to this issue eventually. :)

Comment (chastell) [inline lib/reek/source/source_locator.rb]: Ok, opened https://github.com/troessner/reek/issues/554 to discuss/track this. :)

**Comment by chastell** [comment]
Opened https://github.com/troessner/reek/issues/555 (and assigned to myself) to track the switch to `Pathnames`, merging this one. :)

**Comment by mvz** [comment]
Nice!

**Pull request merged**